### PR TITLE
Add check for Bank of Brazil security module to Windows Flutter Doctor validators

### PR DESCRIPTION
**Pull request by yaakovschectman**
Add a warning to Flutter Doctor if Topaz OFD is found as a process on the system.
The protection module used by the Bank of Brazil has been identified as causing build failures when using VS with CMake for Windows (see https://github.com/flutter/flutter/issues/121366#issuecomment-1845703728). Disabling the software allows the build to succeed again.

If a running process is found by `flutter doctor` whose path contains `Topaz OFD\Warsaw\core.exe`, a warning message is generated to convey this.

Addresses https://github.com/flutter/flutter/issues/121366

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: Could the subprocess fail with an error code? How should that be handled?

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:37]: Is this backwards?

```suggestion
    final String argument = filter != null ? 'Get-Process $filter | Format-List Path' : 'Get-Process | Format-List Path';
```

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: What are your thoughts on returning `partial` with a message indicating that the check failed when the exit code is non-zero?

Comment (zanderso) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: Looking at some of the other examples, I think `missing` might make more sense? Wondering if @christopherfujino has an opinion.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: Yeah, I think missing, because we'd want to investigate further and see if this approach is not robust for all windows setups.

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: Also, please use the ProcessManager class, and not directly invoking the static `Process.run` method

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:8]: if this is a regex, you need to escape the `.` too

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:35]: Looks like this can have a const constructor

**Review by christopherfujino — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: So `missing` when the exit code is nonzero, but is `partial` still desired for when Topaz is found?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:38]: I think missing in both cases. In my mind, partial is just if a validator has sub-validators, with mixed results.

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:54]: let's direct inject this via the constructor. By making this the `LocalProcessManager()` in a local, we could easily accidentally add a "unit" test that ends up spawning real processes on CI.

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:49]: Not anymore now that we inject the process manager.

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:12]: 5 years from now we might forget why this code exists. Could we add a comment that links to https://github.com/flutter/flutter/issues/121366?

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:55]: Could we filter using Dart logic instead and remove the `$filter` argument?

This could be susceptible to command injection attacks if we ever accepted user-controlled input. The code today doesn't seem affected, but I'd avoid this kind of concern altogether if it's not necessary.

Comment (loic-sharma) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:55]: Also, the filter breaks if your machine does not have a process that matches the filter:

image

Standard error:

```
Get-Process : Cannot find a process with the name "core". Verify the process name and call the cmdlet again.
At line:1 char:1
+ Get-Process core
+ ~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (core:String) [Get-Process], ProcessCommandException
    + FullyQualifiedErrorId : NoProcessFoundForGivenName,Microsoft.PowerShell.Commands.GetProcessCommand
```

**Comment by loic-sharma** [comment]
Here's what `flutter doctor` looks like after this change:

image

I suspect the majority of our users don't use Topaz OFD. Could we hide `Topaz OFD` if the validator passes? I'd be fine to include this in the output if `--verbose` option was passed in.

**Comment by christopherfujino** [comment]
> Here's what `flutter doctor` looks like after this change:
> 
> image
> 
> I suspect the majority of our users don't use Topaz OFD. Could we hide `Topaz OFD` if the validator passes? I'd be fine to include this in the output if `--verbose` option was passed in.

We could nest this under Windows, right?

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:49]: fair :)

**Comment by loic-sharma** [comment]
@christopherfujino Nesting is a good idea, but I'd still prefer to hide this unless there's an error if that's possible.

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:12]: Maybe we should call this MalwareScannerValidator, so that we can extend it in the future?

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:22]:
```suggestion
      return const ValidationResult(
```

Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart:22]: oops, put a comment to make this const, but then saw you interpolate exitCode

**Comment by christopherfujino** [comment]
> @christopherfujino Nesting is a good idea, but I'd still prefer to hide this unless there's an error if that's possible.

anything is possible :)

**Comment by yaakovschectman** [comment]
> I suspect the majority of our users don't use Topaz OFD. Could we hide `Topaz OFD` if the validator passes? I'd be fine to include this in the output if `--verbose` option was passed in.

Do we currently have a way to hide passing validators from the output, or would this require updating the validator/doctor framework itself?

**Comment by jonahwilliams** [comment]
From my understanding of the bug, the flutter tool is a third party to the software involved (CMake and this security software). Changes to either piece of software will happen without our involvement might and may invalidate this work and force us to make occassional but disruptive updates (not unlike changes to things like Android Studio validators).

Is the severity of this issue such that we're willing to indefinitely maintain this validation logic that we can't actually test (not counting unit tests)?

**Comment by yaakovschectman** [comment]
> Is the severity of this issue such that we're willing to indefinitely maintain this validation logic that we can't actually test (not counting unit tests)?

The issue is such that we have had multiple Brazilian users unable to build their Windows apps, up until recently, with no known course of action to correct it. In my opinion, making sure residents of one of the world's 7 most populous countries can use Flutter is of considerable importance.

On another note, do we have any existing validators that only output upon error?

> We could nest this under Windows, right?

By `Windows`, do you mean the Windows version validator?

**Comment by loic-sharma** [comment]
> > We could nest this under Windows, right?
> 
> By `Windows`, do you mean the Windows version validator?

Yes that's right

> Is the severity of this issue such that we're willing to indefinitely maintain this validation logic

@jonahwilliams See https://github.com/flutter/flutter/issues/121366, this bug was opened for ~1 year and has 79 comments from 25 participants before we figured it out. There's non-trivial user impact and the workaround is non-obvious. In my mind, `flutter doctor` is the perfect way to raise awareness of this problem.

>  Changes to either piece of software will happen without our involvement might and may invalidate this work and force us to make occassional but disruptive updates (not unlike changes to things like Android Studio validators).

This is a fair concern. Potential risks if CMake/Topaz OFD change:

1. `flutter doctor` has a false positive and incorrectly warns because...
    a. Topaz OFD no longer breaks Flutter Windows build
    b. Some other program was incorrectly detected as Topaz OFD 
2. `flutter doctor` has a false negative and does not warn because...
    a. Topaz OFD changes their process name and is no longer matched by our validator

In my mind, none of these scenarios require a disruptive update (though of course we'd want to fix these ASAP). Does that address your concerns?

**Comment by yaakovschectman** [comment]
> Yes that's right

So what we can do is have the Windows version validator return `missing` with a `ValidationMessage` specific to this issue if `core.exe` is detected, and otherwise nothing new is output. Does this sound like a correct interpretation?

**Comment by jonahwilliams** [comment]
> This is a fair concern. Potential risks if CMake/Topaz OFD change:

3. there is a bug in the validator and/or some change to to cmake or topaz cause the flutter tool or windows build to break.

**Comment by yaakovschectman** [comment]
I've nested the check for Topaz into the `WindowsVersionValidator`. My current approach for integrating the two is when the existing version validator passes, we check for the problematic process. If it is not found, nothing is changed. If it is found, or if `Get-Process` fails, the previous `statusInfo` is added as a `ValidationMessage`, the `type` of the result is changed to `partial`, and `statusInfo` is updated with a message indicating the problem.

File: packages/flutter_tools/lib/src/doctor.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/doctor.dart:147]:
```suggestion
          processLister: globals.processManager,
```

**Comment by christopherfujino** [comment]
As @eliasyishak is already a reviewer on this PR, I will defer to his judgment as far as code review.

File: packages/flutter_tools/lib/src/doctor.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/doctor.dart:147]:
```suggestion
          processLister: ProcessLister(globals.processManager),
```

File: packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart]: Can you remove the changes request if this is resolved?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/topaz_ofd_validator.dart]: As I recall, the only way in GitHub to undo a "requested changes" state is to approve. I re-requested review, then removed myself as a reviewer, and it still shows I requested changes. I will approve once Elias does.

**Comment by yaakovschectman** [comment]
@eliasyishak Do you have any remaining concerns that are not addressed?

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:70]: Is this logic correct? Don't we want to set the `statusInfo` before we add it to the messages list?

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:71]:
```suggestion
          messages.add(const ValidationMessage.hint(
            'The Topaz OFD Security Module process has been found running. '
            'If you are unable to build, you will need to disable it.'
          ));
```

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:48]: Can you also leave this `final` and use the conditionals below to only assign it once

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:69]: Hmm actually the better question would be why we would need 2 messages that are saying the same thing, maybe we remove the first message you are adding and keep the hint validation message below which seems to be more useful

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:69]: Ah this `statusInfo` is holding information saying `Installed version of Windows is version 10 or higher`, my mistake.

File: packages/flutter_tools/test/general.shard/windows_version_validator_test.dart
Comment (eliasyishak) [inline packages/flutter_tools/test/general.shard/windows_version_validator_test.dart:44]: Referring to the style guide, please UpperCamelCase for Identifiers

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:69]: Are you suggesting the `Installed version of Windows is version 10 or higher` should have type `hint`?

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:70]: As you commented below, the purpose here is to create a validation message containing the info regarding the installed Windows version, as that was `statusInfo` before the Topaz related issue was discovered.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:68]: No you are correct, it was not obvious by just looking at the diff that the `statusInfo` variable was already assigned earlier on.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:60]: One more change... can we move all of the logic into a private method for this validator instead actually and have it return a bool, or some other indicator that we can use conditionals against to edit the `statusInfo` and `messages` list?

Doing that will make it easier if we have to add more windows checks in the future and also makes it easier to understand the logic.

You can invoke this method once we have confirmed that the windows version is valid

Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:60]: Would we actually want a `bool` out of that? It seems to me that we would also want the status info and/or validation messages list to depend on this logic, and they may have more than just 2 possible states.

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:60]: Yeah agreed, the `bool` was just an example of what can be returned. You can return a pair of values, one for status and another for messages that get joined together.

This is to just make it so that we can maintain this validator easily in the future and possibly add more if needed

Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:60]: It sounds like this private method would then just essentially return a `ValidationResult`. What logic are you referring to moving? Just the version check, the processes check, or both? If both, would not `validate` just turn into a wrapper around calling this private method? I am not sure I understand what change you are proposing.

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:60]: Yeah, sorry if it's not coming off clear, the below is what I was thinking

```dart
    // Use the string split method to extract the major version
    // and check against the [kUnsupportedVersions] list
    ValidationType windowsVersionStatus;
    String statusInfo;
    final List<ValidationMessage> messages = <ValidationMessage>[];
    if (matches.length == 1 && !kUnsupportedVersions.contains(matches.elementAt(0).group(1))) {
      windowsVersionStatus = ValidationType.success;
      statusInfo = 'Installed version of Windows is version 10 or higher';
      messages.add(ValidationMessage(statusInfo));

      // List of windows specific checks which can be unrelated to versions but
      // could be causing users problems
      final List<ValidationResult> windowsSubResults = [
        _topazFinder(), // The private method with the logic needed
      ];
      for (final result in windowsSubResults) {
        if (result.type != ValidationType.success) {
          statusInfo = 'Problem detected with Windows installation';  // This should be more generic while we list the problem in the messages
          windowsVersionStatus = Validation.partial;
          messages.add(result.message);
        }
      }

    } else {
      windowsVersionStatus = ValidationType.missing;
      statusInfo = 'Unable to determine Windows version (command `ver` returned $commandResult)';
    }
```

This way if we need to add more checks like yours in the future it should be easy for us to add it to the list of checks

Also, for this to work, I walk back my comment about making the return variables `final`.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:46]: Would it make sense to do a string contains instead of a regex match?

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:48]: Consider formatting this into multiple lines to improve readability. Something like:

```suggestion
        return const ValidationResult(
          ValidationType.missing,
          <ValidationMessage>[
            ValidationMessage.hint(
              'The Topaz OFD Security Module process has been found running. '
              'If you are unable to build, you will need to disable it.',
            ),
          ],
        );
```

**Review by eliasyishak — APPROVED** [review]
LGTM after @loic-sharma's comments are finished! Thank you for refactoring

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:46]: I suppose we could pass it a slightly more convoluted pattern in order to use that method. I think it is best that we make the test case-insensitive and perhaps should use a regex shorthand for the space character.

Comment (loic-sharma) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:46]: Since this contains only ASCII characters, you can use lowered strings to do a case-insensitive string comparison. Matching only the ASCII space (`U+0020`) seems acceptable.

Feel free to upvote this Dart request for `String.equalsIgnoreCase`: https://github.com/dart-lang/sdk/issues/54418 😅

Comment (yaakovschectman) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:46]: Do you have any other concerns on this change?

**Comment by yaakovschectman** [comment]
@christopherfujino With Elias' approval submitted, do you have any remaining concerns or requests blocking this merge?

**Review by christopherfujino — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:53]: What do you think of this message?

```suggestion
              'The Topaz OFD Security Module was detected on your machine. '
              'You may need to disable it to build Flutter applications.',
```

**Review by loic-sharma — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
